### PR TITLE
Refactor command line parsing to retrieve Embulk System Properties in the very beginning

### PR DESCRIPTION
**Pull request by dmikurube**
To realize the goals below, this PR refactors the entire command line processing:

* Embulk CLI can retrieve "Embulk System Properties" values from the command-line before starting the actual Embulk processing.
* The "Embulk System Properties" are available for sub commands like `embulk gem` and `embulk bundle`.
    * We are going to need this to use JRuby out of the Embulk executable binary -- whose location is specified by `jruby` Embulk system property.

It refactors `EmbulkRun` (the main entry logic when Embulk runs from CLI, not `EmbulkEmbed`) as well because the command line processing and the main entry logic have been mixed so far. `EmbulkRun` has got much simpler.

----

JFYI, Embulk is using Apache Commons-CLI, a kind of old-fashioned command line processor library for some reasons.

* Originally (before Embulk v0.8.X), Embulk's command line processing was implemented in Ruby. When we reimplemented all CLI processing in Java, we chose Apache Commons-CLI because it was the only one that was able to implement compatible command line syntax at that time.
* Apache Commons-CLI is old-fashioned, but known to be stable. We hardly have to change the command line parsing by updates.
* ...

### CODE REVIEW

File: embulk-core/src/main/java/org/embulk/cli/Command.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    public static Command of(final String command) {
+        if (command == null) {
+            return NONE;
+        }
+
+        for (final Command constant : values()) {
+            if (constant == UNKNOWN || constant == NONE) {
+                continue;
+            }
+            if (constant.toString().equals(command)) {
```
[the remark this answers is not in the document]
Comment (dmikurube) [reply embulk-core/src/main/java/org/embulk/cli/Command.java:34]: `constant.toString()` is based on `this.command` initialized with lower-case string literals above. (It's not from the automated translation by Java enums.)

At the same time, the argument `command` is from user's input (command line), and Embulk has not accepted upper-case commands so far.

Then, I thought it's okay. If you think we shouldn't use this kind of redundant `this.command` argument, or we should accept upper-case commands, please let me know. (I think that would be a valid suggestion.)

[the remark this answers is not in the document]
Comment (shroman) [reply embulk-core/src/main/java/org/embulk/cli/Command.java:34]: No strong preference really. With "Embulk has not accepted upper-case commands" all looks good, especially for CLI :)

**Review by shroman — APPROVED** [review]
LGTM

File: embulk-core/src/main/java/org/embulk/cli/CommandLine.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
+package org.embulk.cli;
+
+import java.util.List;
+import java.util.Properties;
+
+public abstract class CommandLine {
```
Comment (trung-huynh) [inline embulk-core/src/main/java/org/embulk/cli/CommandLine.java:6]: curious, why not this one an interface instead of abstract class?

**Review by trung-huynh — APPROVED** [review]
LGTM

File: embulk-core/src/main/java/org/embulk/cli/CommandLine.java (hunk 2 above)
Comment (dmikurube) [reply embulk-core/src/main/java/org/embulk/cli/CommandLine.java:6]: No strong reasons. 
* It had some implemented methods when I started tries-and-errors on this, but they finally got away.
* This is just a stub of `embulk-deps`' `CommandLineImpl`, not expecting to have various classes implementing the `CommandLine` interface.

**Comment by dmikurube** [comment]
Thanks! Merging this and #1334, and then releasing v0.10.20.

**Pull request merged**